[clang][Sema] Fix false positive -Wshadow with structured binding captures

ivanmurashko opened the pull request:
Previously, lambda init captures of structured bindings were incorrectly classified as regular shadow warnings (shown with `-Wshadow`), while regular parameter captures were correctly classified as `uncaptured-local` warnings (shown only with `-Wshadow-all`). This created inconsistent behavior:

```cpp
  void foo1(std::pair<int, int> val) {
    [val = std::move(val)](){}();  // No warning with -Wshadow (correct)
  }

  void foo2(std::pair<int, int> val) {
    auto [a, b] = val;
    [a = std::move(a)](){}();      // Warning with -Wshadow (incorrect)
  }
```

The fix extends the existing lambda capture classification logic in `CheckShadow()` to handle `BindingDecl` consistently with `VarDecl`, ensuring both cases show no warnings with `-Wshadow` and `uncaptured-local` warnings with `-Wshadow-all`.

Fixes #68605.

ivanmurashko committed:
[clang][Sema] Fix false positive -Wshadow with structured binding captures

Previously, lambda init captures that captured structured bindings would
incorrectly emit shadow warnings, even though regular parameter captures
don't emit such warnings. This created inconsistent behavior:

```cpp
void foo1(std::pair<int, int> val) {
  [val = val](){}();  // No warning (correct)
}

void foo2(std::pair<int, int> val) {
  auto [a, b] = val;
  [a = a](){}();      // Warning (incorrect)
}
```

The fix modifies getShadowedDeclaration() for VarDecl to return nullptr
when a lambda init capture would shadow a BindingDecl, ensuring consistent
behavior between regular captures and structured binding captures.

zwuis commented:
Can you tell why there is false positive with structured bindings only, but without variables?

ivanmurashko commented:
> Can you tell why there is false positive with structured bindings only, but without variables?

Structured bindings (`auto [a, b] = pair`) create `BindingDecl` AST nodes, while regular variables create `VarDecl` nodes. Both actually trigger shadow warnings in lambda captures, but they get classified into different warning categories:
  - Regular variables: `VarDecl` shadowing `VarDecl -> classified as `-Wshadow-uncaptured-local` (only shown with `-Wshadow-all`)
  - Structured bindings: `VarDecl` shadowing `BindingDecl` -> classified as `-Wshadow` (shown with basic `-Wshadow`)

This makes regular variables appear to "not warn" in typical usage (since most users use -Wshadow, not -Wshadow-all), while structured bindings always warn with basic -Wshadow. The "false positive" occurs because users expect consistent behavior between these equivalent patterns:
```
  [val = val]  // No warning with -Wshadow (classified as uncaptured-local)
  [a = a]      // Warning with -Wshadow (classified as regular shadow) 
```

My fix was focused on the simplest solution to make behavior consistent for the most popular case (basic `-Wshadow`). A proper fix should probably preserve the `uncaptured-local` warning classification for structured bindings rather than  suppressing all shadow warnings, but that would require more complex changes to the warning classification logic.

ivanmurashko committed:
[Sema] Fix inconsistent shadow warnings for lambda capture of structured bindings

Lambda captures that shadow structured bindings were incorrectly classified
as regular shadow warnings (shown with -Wshadow) while regular parameter
captures were classified as uncaptured-local warnings (shown only with
-Wshadow-all). This created inconsistent behavior between semantically
equivalent code patterns.

This change extends the existing lambda capture classification logic to
handle BindingDecl consistently with VarDecl:
- Lambda init captures of structured bindings now show as uncaptured-local
- Regular variable declarations inside lambda bodies still show as shadow
- All existing shadow warning functionality is preserved

The fix ensures consistent behavior between:
  void func(std::pair<int,int> val) { [val = val](){}; }     // no -Wshadow, warns with -Wshadow-all
  void func(std::pair<int,int> val) { auto [a,b] = val; [a = a](){}; }  // no -Wshadow, warns with -Wshadow-all

Previously the structured binding case incorrectly produced warnings with
basic -Wshadow, preventing structured bindings from being used in lambda
captures consistently with regular parameters.

ivanmurashko commented:
> So current fix will introduce false negative with "-Wshadow-uncaptured-local" and shadowing structured bindings, and, **hide the underlying issue**. It would be great to find the difference of handling shadowing `VarDecl` and `BindingDecl`.

The new commit c0723fefc075 addresses this by extending the existing lambda capture classification logic at lines 8501-8515 in `clang/lib/Sema/SemaDecl.cpp`. 

Previously, `VarDecl` had lambda capture handling (lines 8484-8500) while `BindingDecl` fell through to regular shadow detection, causing inconsistent warning classification. The fix ensures both `VarDecl` and `BindingDecl` follow the same lambda capture code path, making `[val = val]` and `[a = a]` (where `a` is from structured binding) behave identically.

ivanmurashko committed:
Fix warn-shadow-in-lambdas test for structured bindings

zwuis reviewed (commented):
Thanks! It looks better.

Please add a release note entry to "clang/docs/ReleaseNotes.rst".

ivanmurashko committed:
[Sema] Consolidate VarDecl and BindingDecl shadow detection using ValueDecl

Consolidate the handling of VarDecl and BindingDecl in shadow detection
by using their common base class ValueDecl.

ivanmurashko commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -8410,6 +8410,7 @@ NamedDecl *Sema::getShadowedDeclaration(const VarDecl *D,
     return nullptr;
 
   NamedDecl *ShadowedDecl = R.getFoundDecl();
+
```
> Seems unrelated.

Restored at 27fc9b7095e6ab993782e121e1ff1a3776b2ab0a

ivanmurashko commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -8496,6 +8498,21 @@ void Sema::CheckShadow(NamedDecl *D, NamedDecl *ShadowedDecl,
+              return;
+            }
+          }
+          // For non-init-capture cases, fall through to regular shadow logic
```
Consolidated `VarDecl` and `BindingDecl` handling using `ValueDecl` at 27fc9b7095e6ab993782e121e1ff1a3776b2ab0a

ivanmurashko commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -8508,10 +8525,17 @@ void Sema::CheckShadow(NamedDecl *D, NamedDecl *ShadowedDecl,
+      // Apply scoping logic to both VarDecl and BindingDecl
+      bool shouldApplyScopingLogic = false;
+      if (const auto *VD = dyn_cast<VarDecl>(ShadowedDecl)) {
+        shouldApplyScopingLogic = VD->hasLocalStorage();
```
Done at 27fc9b7095e6ab993782e121e1ff1a3776b2ab0a

ivanmurashko replied:
Addressed at 578e61b7aebbfa9213a28f3dabd29708112ce2a5 (also changed local vars names according the coding style)

ivanmurashko committed:
[Clang] Simplify BindingDecl scoping logic using getDecomposedDecl()->hasLocalStorage()

ivanmurashko committed:
[Clang] Add release note for consistent structured binding shadow warnings

Fznamznon reviewed (approved):
Just style nits, otherwise LGTM

ivanmurashko committed:
Apply suggestions from code review

Code style fixes applied

Co-authored-by: Mariya Podchishchaeva <mariya.podchishchaeva@intel.com>

ivanmurashko committed:
[Clang] Add missing newline at end of PR68605.cpp test file

ivanmurashko commented:
I'll wait until Monday to provide additional time for review. If there are no further comments by then, I'll proceed with merging the PR.

ivanmurashko committed:
Merge branch 'main' into fix-structured-binding-lambda-shadow-68605
